core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/udf/PythonUDFOperator.java

```diff
@@ -0,0 +1,700 @@
+            for (String s : outerFileNames) {
+                outerFilePaths.add(Utils.getResourcePath(s, TexeraProject.TEXERA_DATAFLOW).toString());
+            }
+        } else outerFilePaths = null;
+        mapper = new ObjectMapper();
+    }
+
+    public void setInputOperator(IOperator operator) {
+        if (cursor != CLOSED) {
+            throw new TexeraException("Cannot link this operator to another operator after the operator is opened");
+        }
+        this.inputOperator = operator;
+    }
+
+    /*
+     * add new field(s) to the schema, with name resultAttributeName and type TEXT
+     */
+    private Schema transformSchema(Schema inputSchema){
+        for (String name :  predicate.getInputAttributeNames()) {
+            Schema.checkAttributeExists(inputSchema, name);
+        }
+        for (String name : predicate.getResultAttributeNames()) {
+            Schema.checkAttributeNotExists(inputSchema, name);
+        }
+        Schema transformedSchema = convertToTexeraSchema(convertToArrowSchema(inputSchema));
+        Schema.Builder resultSchemaBuilder = new Schema.Builder().add(transformedSchema);
+        List<String> resultNames = predicate.getResultAttributeNames();
+        List<AttributeType> resultTypes = predicate.getResultAttributeTypes();
+        for (int i = 0; i < resultNames.size(); i++) {
+            resultSchemaBuilder.add(resultNames.get(i), resultTypes.get(i));
+        }
+        return resultSchemaBuilder.build();
+    }
+
+    /**
+     * When this operator is opened, it executes the python script, which constructs a {@code FlightServer}
+     * object which is then up and running in the specified address. The operator calls
+     * {@code flightClient.doAction(new Action("healthcheck"))} to check the status of the server, and then proceeds if
+     * successful (otherwise there will be an exception).
+     * @throws TexeraException
+     */
+    @Override
+    public void open() throws TexeraException {
+        if (cursor != CLOSED) {
+            return;
+        }
+        if (inputOperator == null) {
+            throw new DataflowException(ErrorMessages.INPUT_OPERATOR_NOT_SPECIFIED);
+        }
+
+        inputOperator.open();
+
+        // Flight related
+        try {
+            int portNumber = getFreeLocalPort();
+            Location location = new Location(URI.create("grpc+tcp://localhost:" + portNumber));
+            List<String> args = new ArrayList<>(
+                    Arrays.asList(PYTHON, PYTHONSCRIPT, Integer.toString(portNumber), userScriptPath)
+            );
+
+            ProcessBuilder processBuilder = new ProcessBuilder(args).inheritIO();
+            // Start Flight server (Python process)
+            processBuilder.start();
+            // Connect to server
+            boolean connected = false;
+            int tryCount = 0;
+            while (!connected && tryCount < 5) {
+                try {
+                    Thread.sleep(2000);
```

Comment:
I disagree with `Thread.sleep` here. This means for any operator, this operation would take at least 2 seconds. This would be way too slow. Furthermore, I checked `flichtClient.doAction` function, it implementation is already a blocking call `blockingServerStreamingCall`.  If we remove the `Thread.sleep`, what will happen? What will the `message` be?

Reply:
`doAction` is a blocking call, but before that, when `client` starts and connect to the server in `FlightClient.builder(globalRootAllocator, location).build()`, this is a non-blocking call. In my machine (Windows), not adding a sleep before this will not be a problem, but it seems in a Unix-based OS, this `build()` will have a much smaller delay, and this delay cannot be set manually. Since the server is also started in `open()`, and without connecting to the server there's no way to know the status, here a `sleep` will have to be added, otherwise for scripts that may take longer to start the try count will be exceeded very fast on a Unix-based OS. However, what we could do is reduce this delay time and increase the try count.

Reply:
Changes are done according to what I suggested.